doc/monitoring_containers/monitoring_containers.md

```diff
@@ -0,0 +1,271 @@
+Monit will periodically read the target processes from configuration file and try to match 
+those process with the processes tree in Linux kernel.
+
+Below is an example of Monit configuration file to monitor the critical processes in lldp
+container.
+
+*/etc/monit/conf.d/monit_lldp*
+```bash
+###############################################################################
+# Monit configuration file for lldp container
+# Process list:
+#   lldpd
+#   lldp_syncd
+#   lldpmgrd
+###############################################################################
+check process lldp_monitor matching "lldpd: "
+    if does not exit for 5 times within 5 cycles then alert
+check process lldp_syncd matching "python2 -m lldp_syncd"
+    if does not exit for 5 times within 5 cycles then alert
+check process lldpmgrd matching "python /usr/bin/lldpmgrd"
+    if does not exit for 5 times within 5 cycles then alert
+```
+
+### 2.2.2 Monitoring Critical Resource Usage
+Similar to monitoring the critical processes, we can employ Monit to monitor the resource usage
+such as CPU, memory and disk for each process. Unfortunately Monit is unable to do the resource monitoring
+in the container level. Thus we propose a new design to achieve such monitoring based on Monit.
+Specifically Monit will monitor a script and check its exit status. This script
+will correspondingly read the resource usage of docker containers, compare it with
+pre-defined threshold and then return a value. The value 0 signified that
+the resource usage is less than threshold and non-zero means Monit will send an alert since
+current usage is larger than threshold.
+
+Below is an example of Monit configuration file for lldp container to pass the pre-defined 
+threshold (bytes) to the script and check the exiting value.
+
+```bash
+check program container_memory_lldp with path "/usr/bin/memory_checker lldp 104857600"
+    if status != 0 then alert
+```
+
+### 2.2.3 Auto-restart Docker Container
+The design principle behind this auto-restart feature is docker containers can be automatically shut down and
+restarted if one of critical processes running in the container exits unexpectedly. Restarting
+the entire container ensures that configuration is reloaded and all processes in the container
+get restarted, thus increasing the likelihood of entering a healthy state.
+
+Currently SONiC used supervisord system tool to manage the processes in each
+docker container. Actually auto-restarting docker container is based on the process 
+monitoring/notification framework. Specifically
+if the state of process changes for example from running to exited,
+an event notification `PROCESS_STATE_STOPPED` will be emitted by supervisord.
+This event will be received by event listener. If the exited process is critical
+one, then the event listener will terminate supervisord and the container will be shut down
+and restarted.
```

Comment:
```
This event will be received by event listener. The event listener determines if the process is a critical
process and whether it exited unexpectedly. If both of these conditions are true, the event listener will kill the
supervisord process. Since supervisord runs as PID 1 inside the containers, when supervisord exits, the container will
stop. When the container stops, the systemd service which manages the container will also stop, but it is configured
to automatically restart the service, thus it will restart the container.
```

Reply:
Reworded.